File: src/main/scala/cn/cjie/scala/akka/typedactor/Calculator.scala
package cn.cjie.scala.akka.typedactor

import java.util.concurrent.TimeUnit

import akka.actor.{OneForOneStrategy, SupervisorStrategy, ActorRef, TypedActor}
import akka.util.Timeout
import scala.concurrent.{Promise, Future}
import akka.actor.TypedActor.{PreStart, PostStop}
import akka.event.Logging
import akka.actor.SupervisorStrategy.{Escalate, Stop, Restart, Resume}

/**
 * Created with IntelliJ IDEA.
 * User: hucj
 * Date: 14-11-4
 * Time: 上午9:55
 * To change this template use File | Settings | File Templates.
 */
class Calculator extends CalculatorInt with PreStart with PostStop {
  var counter: Int = 0
  import TypedActor.dispatcher
  def add(first: Int, second: Int): Promise[Int] =
    Promise.successful(first + second)
  def subtract(first: Int, second: Int): Promise[Int] =
    Promise successful first - second
  def incrementCount(): Unit = counter += 1
  def incrementAndReturn(): Option[Int] = {
    counter += 1
    Some(counter)
  }
  import  TypedActor.context
  val log = Logging(context.system,
    TypedActor.self.getClass())
  def preStart(): Unit = {
    log.info ("Actor Started")
  }
  def postStop(): Unit = {
    log.info ("Actor Stopped")
  }

  def onReceive(message: Any, sender: ActorRef): Unit = {
    log.info("Message received->{}", message)
  }

  def supervisorStrategy(): SupervisorStrategy =
    OneForOneStrategy(maxNrOfRetries = 10,
      withinTimeRange = Timeout(10, TimeUnit.SECONDS).duration) {
      case _: ArithmeticException => Resume
      case _: IllegalArgumentException => Restart
      case _: NullPointerException => Stop
      case _: Exception => Escalate
    }
}

File: src/main/java/com/cjie/akka/typedactor/Calculator.java
package com.cjie.akka.typedactor;

import akka.actor.ActorRef;
import akka.actor.OneForOneStrategy;
import akka.actor.SupervisorStrategy;
import akka.actor.TypedActor;
import akka.dispatch.Futures;
import akka.event.Logging;
import akka.event.LoggingAdapter;
import akka.japi.Function;
import akka.japi.Option;
import akka.util.Timeout;
import scala.concurrent.Future;

import java.util.concurrent.TimeUnit;

/**
 * Created with IntelliJ IDEA.
 * User: hucj
 * Date: 14-11-4
 * Time: 上午9:49
 * To change this template use File | Settings | File Templates.
 */
public class Calculator implements CalculatorInt, TypedActor.PreStart, TypedActor.PostStop {
    Integer counter = 0;
    //Non-blocking request response
    public Future<Integer> add(Integer first, Integer second) {
        return Futures.successful(first + second);
    }
    //Non-blocking request response
    public Future<Integer> subtract(Integer first,
                                    Integer second) {
        return Futures.successful(first - second);
    }
    //fire and forget
    public void incrementCount() {
        counter++;
    }
    //Blocking request response
    public Option<Integer> incrementAndReturn() {
        return Option.some(++counter);
    }

    LoggingAdapter log = Logging.getLogger(
            TypedActor.context().system(), this);
    //Allows to tap into the Actor PreStart hook
    public void preStart() {
        log.info("Actor Started !");
    }
    //Allows to tap into the Actor PostStop hook
    public void postStop() {
        log.info("Actor Stopped !");
    }

    public void onReceive(Object msg, ActorRef actor) {
        log.info("Received Message -> {}", msg);
    }

    public SupervisorStrategy supervisorStrategy() {
        return strategy;
    }
    private static SupervisorStrategy strategy =
            new OneForOneStrategy(10, new Timeout(10, TimeUnit.SECONDS).duration(),
                    new Function<Throwable, SupervisorStrategy.Directive>() {
                        public SupervisorStrategy.Directive apply(Throwable t) {
//                            if (t instanceof ArithmeticException) {
//                                return resume();
//                            } else if (t instanceof IllegalArgumentException) {
//                                return restart();
//                            } else if (t instanceof NullPointerException) {
//                                return stop();
//                            } else {
//                                return escalate();
//                            }
                            return  null;
                        }
                    });
}
